includes/panopto_derivatives.inc

```diff
@@ -5,37 +5,386 @@
+        watchdog_exception('islandora_video_streaming_panopto', $ex);
+      }
+    } else {
+      $messages = array(
+        'message' => t('Failed to retrieve session for object: @pid'),
         'message_sub' => array('@pid' => $object->id),
         'type' => 'watchdog',
         'severity' => WATCHDOG_ERROR,
-      ),
-    );
-
-    //TODO: upload, record result in $streaming
- 
+      );
+      $failureMessage['messages'][] = $messages;
+    }
+    file_delete($archival_file['file'], true);
     if ($streaming) {
        // Write the data to a temporary file.
       $temp_name = drupal_tempnam('temporary://', 'file');
       if (file_put_contents($temp_name, json_encode($streaming)) === FALSE) {
-        $failureMessage['messages'][0]['message'] = t('Failed to create temporary file @filename for object @pid when uploading to Panopto.');
-        $failureMessage['messages'][0]['message_sub']['@filename'] = $temp_name;
+        $messages = array(
+            'message' => t("Failed to create temporary file @filename for object @pid when uploading to Panopto"),
+            'message_sub' => array(
+              '@pid' => $object->id, 
+              '@filename' => $temp_name
+            ),
+            'type' => 'watchdog',
+            'severity' => WATCHDOG_ERROR,
+        );
+        $failureMessage['messages'][] = $messages;
         return $failureMessage;
       }
-      return islandora_video_add_datastream($object, 'KALTURA_STREAM', $temp_name);
+      return islandora_video_add_datastream($object, 'PANOPTO_INFO', $temp_name);
     } else {
       return $failureMessage;
     }
   }
 }
 
+/**
+ * Requests token and returns token
+ * @param $oauthendpoint string The authentication URL for request
+ * @return string A token retrieved from server
+ */
+function islandora_video_streaming_panopto_get_token($oauthendpoint) {
+  $client_id = variable_get('islandora_video_streaming_panopto_client_id');
+  $secret_client_id = variable_get('islandora_video_streaming_panopto_secret_client_id');
+  $username = variable_get('islandora_video_streaming_panopto_username');
+  $password = variable_get('islandora_video_streaming_panopto_password');
+  
+  $response = null;
+  $token = null;
+  
+  if (isset($oauthendpoint) && isset($client_id) && isset($secret_client_id) && isset($username) && isset($password)) {
+    $authorization = base64_encode("$client_id:$secret_client_id");
+    $headers = array("Authorization: Basic {$authorization}", "Content-Type: application/x-www-form-urlencoded");
+    $content = 'grant_type=password&username='.urlencode(strtolower($username)).'&password='.urlencode($password).'&scope=api';
+    $response = islandora_video_streaming_panopto_request_curl($oauthendpoint, $headers, $content, 'POST', array(200 => 'access_token'));
+  }
+  if ($response) {
+    $token = $response['access_token'];
+  }
+  return $token;
+}
+
+/**
+ * Retrieves session needed for upload
+ * @param $token string The token retrieved from server
```

Comment:
Here and elsewhere, you need to reverse the order of `type` and `name`.  https://docs.phpdoc.org/latest/references/phpdoc/tags/param.html